### ISSUE TITLE
two questions/issues with Alyx.newExp

### ISSUE THREAD

**Issue by nsteinme** [cortex-lab/alyx-matlab#74]
Hi @k1o0 - two things about Alyx.newExp: 
1) Why does it give an error if the subject does not come up in dat.listSubjects (which is to say, if the subject does not have a folder on the server), rather than simply creating the folder at that time? If the subject does exist in Alyx (which is the check that actually should happen here) then the folder can just be created now. This would also allow us to NOT create folders from eui.AlyxPanel, which pollutes the server directory tree with folders for mice who will never have experiments run. 

2) I don't think the part of the code where it is supposed to find and re-use the "Base" session is working. We just get a new Base session for every single experiment. Do you see instead that there's only one base session with subsequent experimental sessions? 
![image](https://user-images.githubusercontent.com/7590975/88580873-29059d80-d001-11ea-88b9-717a2850a509.png)

Let me know what you think about these things, then maybe we can make separate issues for different parts of the solutions.

**Comment by k1o0** [comment]
I'll take a closer look this evening but here are my first thoughts:

1. Alyx should be using it's own `listSubjects` method, and each time you log in, Alyx should create all the missing subject folders.  If you created the subject on Alyx `after` logging in to the Alyx class then it wouldn't have created the local folder.  We can add a case for this but it's never come up before because we typically add subjects long before we create a session.

2. I suspect that the REST API has changed a bit and that's causing this issue.  I haven't run the local tests in a while so I'll see what's happening there.

**Comment by nsteinme** [comment]
Thanks for testing # 2. For # 1, agreed alyx should use its own listSubjects. But I do suggest the folder be made at the time of newExp for that subject, rather than at the time of logging in. E.g. right now my server has about 500 folders for every IBL mouse - this is not useful...

**Comment by k1o0** [comment]
>But I do suggest the folder be made at the time of newExp for that subject, rather than at the time of logging in
The reason for creating them at login was because if Alyx went down and mc went into headless mode then the subject folder would most likely already be there, so there wouldn't be the problem of not being able to run the subject.  If the subject folder isn't there (and Alyx is down) then you have to create the folder and restart mc.  

We should probably change the code to only create the subject folders for your lab's mice...

**Comment by k1o0** [comment]
I've run the tests and fixed the issue.  The sessions switched from being returned in ascending order to descending, which confused my code.  I've also added a specific test for base session creation.

The change is on `master` and `2.5.0`.  You can also cherry-pick 6fa3d07.

**Comment by nsteinme** [comment]
Awesome thanks Miles. But can you clarify what you mean about the folder creation - you say "if the subject folder is already there, then there wouldn't be the problem of not being able to run the subject". But I am saying that the subject folder should be created, if it does not exist, at the time of newExp. So you would never fail to run the subject for this reason - in fact this would actually be even better insurance that the folder is present when you try to run, rather than hoping it was created earlier. I think I didn't understand your point clearly?

**Comment by k1o0** [comment]
In the early days people were very uncomfortable with using Alyx with Rigbox, and the database was less stable.  I added a lot of safeguards so that if Alyx failed nothing would be affected.  When logged in the subject drop-down lists all the subjects in Alyx.  Without folder creation if you logged out (or Alyx became inaccessible) the subjects would revert to the directory list on the local server and any mice without a folder would vanish.  Users would then have to manually create the folder on the local server and restart mc which they wouldn't necessarily think to do.  If all the subject folders are created whenever _any_ user logs in then the likelihood of there already being a subject folder on the server becomes much higher.  If Alyx goes down and the user wants to run an experiment, chances are their subject is in the local subject list, even if they've never run a session before. 

Alyx is much more stable now and people can't easily run an experiment without being logged in unless Alyx is inaccessible so this is no longer such a concern.  I think I'll do two things: 1. Add folder creation in `Alyx/newExp` if it's not already there; 2. Add an optional preference in the paths file for lazy folder creation and an optional local lab name override.  I'll add these as separate issues and close this one.

BTW I hope to take a week off IBL next week to concentrate on Rigbox.  This will probably happen in August or September.  If there's anything you feel needs prioritizing during that time let me know.  I'll plan to use that time to update the MC UI and finish off the passive movie presentation experiment class I created.
